src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,2689 @@
+export presentation
+
+abstract type ModuleFP{T} end
+
+const CRing = Union{MPolyRing, MPolyQuo{<:Oscar.MPolyElem}, MPolyRing_dec, MPolyQuo{<:Oscar.MPolyElem_dec}}
+const CRingElem = Union{MPolyElem, MPolyQuoElem{<:Oscar.MPolyElem}, MPolyElem_dec, MPolyQuoElem{<:Oscar.MPolyElem_dec}}
+#TODO: "fix" to allow QuoElem s as well...
+# this requires
+#  re-typeing of FreeModule
+#  typing of BiModArray
+# ... and all the rest.
+# parametrization has to be by elem_type(coeff_ring) and not, like currently, the bottom coeff ring
+# Also: qring is a Singular native. So it needs to be added to the ring creation
+
+abstract type ModuleFPHom end
+abstract type ModuleMap{T1, T2} <: Map{T1, T2, Hecke.HeckeMap, ModuleFPHom} end
+
+
+# For all module types, including free modules we store canonical in and outgoing morphisms 
+# which were generated by the function creating the module
+# otherwise a free module consists out of a rank and a ring
+# one can also provide names for the generators
+mutable struct FreeMod{T} <: ModuleFP{T}
```

Comment:
What is the `T` doing here? It is not u sed insider. So either we don't need it, and can drop it; or we do need it; maybe it is meant to restrict `R`?  If so, it may not be possible to express this via the type system (?), so if not, I suggest the inner constructor(s) perform at least a runtime check to guarantee the desired invariant on the types (whatever it is)